### PR TITLE
Feature: add process_func_applies_sliding_window argument

### LINKED ISSUE #54: Feature: add option to call process function on a sliding window
Currently, when working with a sliding window, the user is responsible to handle the windowing. Unless in the case that a function from an external library is called that already implements a sliding window, this makes it quite inconvenient for a user to work with this feature. E.g. the following (not very efficient) code is needed to just calculate the mean over a sliding window:

```python
def mean(signal, sampling_rate, win_dur, hop_dur):
    win_dur = int(sampling_rate * win_dur)
    hop_dur = int(sampling_rate * hop_dur)
    steps = (signal.shape[-1] - win_dur) // hop_dur
    pos = 0
    y = []
    for _ in range(steps):
        win = signal[0, pos:pos+win_dur]
        y.append(win.mean())
        pos += hop_dur
    return np.atleast_2d(y)

win_dur=1.0
hop_dur=0.5
interface = audinterface.Feature(
    'mean',
    process_func=mean,
    process_func_args={
        'win_dur': win_dur,
        'hop_dur': hop_dur,
    },
    win_dur=win_dur,
    hop_dur=hop_dur,
)
```

And it's also easy to mess up, e.g. the following would work, but produce wrong timestamps:

```python
win_dur=1.0
hop_dur=0.5
interface = audinterface.Feature(
    'mean',
    process_func=mean,
    process_func_args={
        'win_dur': win_dur,
        'hop_dur': hop_dur,
    },
    win_dur=win_dur * 2,
    hop_dur=hop_dur * 2,
)
```

I think it would be very convenient to offer an option that the windowing is handled internally so that the user only has to provide the actual processing code, e.g.:

```
def mean(signal, sampling_rate):
    return signal.mean()

win_dur=1.0
hop_dur=0.5
interface = audinterface.Feature(
    'mean',
    process_func=mean,
    process_func_sliding=True,  # apply sliding window on signal and forward only single frames to processing function
    win_dur=win_dur,
    hop_dur=hop_dur,
)
```

### PR DESCRIPTION
**Pull request by frankenjoe**
Closes #54 

### API

image

### Tutorial

image

image

### CODE REVIEW

File: tests/test_feature.py [hunk 1]
```diff
@@ -910,15 +946,38 @@ def test_signal_sliding_window(win_dur, hop_dur):
...
+
+    interface = audinterface.Feature(
+        feature_names='mean',
+        process_func=mean_sliding_window,
+        process_func_args={
+            'win_dur': 0.5,
+            'hop_dur': 0.25,
+        },
+        process_func_is_mono=False,
+        process_func_applies_sliding_window=True,
```
Comment (hagenw) [inline tests/test_feature.py:967]: Why do we need the `process_func_applies_sliding_window` argument?
We have set the window and hop duration anyway as parameter of the sliding window. Or is it just to add a check to ignore the `win_dur` and `hop_dur` arguments inside `Feature`?

Comment (frankenjoe) [reply tests/test_feature.py:967]: We need some way to tell `Feature` whether the process function will apply the sliding window or if this should be handled internally. So do you propose we should derive that from the arguments to the processing function?

Comment (frankenjoe) [reply tests/test_feature.py:961]: The other problem is that even if the processing function does not apply the sliding window itself, it may still need the information about the window and hop duration. In that case it would be a bad idea to automatically derive the strategy from the fact that the processing function has a `win_dur` and `hop_dur` argument. So guess it's safer to introduce an argument to control the behavior.

However, I now implemented that the `win_dur` and `hop_dur` arguments are automatically added to `process_func_args` if the processing function expects such arguments and they are not explicitly set by the user. Now:

```python
db = audb.load(
    'emodb',
    version='1.2.0',
    format='wav',
    mixdown=True,
    sampling_rate=16000,
    full_path=False,
)


def mean(signal, sampling_rate, win_dur, hop_dur):
    frames = audinterface.utils.sliding_window(
        signal,
        sampling_rate,
        win_dur,
        hop_dur,
    )
    return frames.mean(axis=1, keepdims=True)


win_dur=1.0
hop_dur=0.5

interface = audinterface.Feature(
    'mean',
    process_func=mean,
    process_func_applies_sliding_window=True,
    process_func_args={
        'win_dur': win_dur,
        'hop_dur': hop_dur,
    },
    win_dur=win_dur,
    hop_dur=hop_dur,
)
interface.process_files(db.files[:3], root=db.root)
```
```
                                                                   mean
file            start                  end                             
wav/03a01Fa.wav 0 days 00:00:00        0 days 00:00:01        -0.000329
                0 days 00:00:00.500000 0 days 00:00:01.500000 -0.000285
wav/03a01Nc.wav 0 days 00:00:00        0 days 00:00:01         0.000039
                0 days 00:00:00.500000 0 days 00:00:01.500000 -0.000412
wav/03a01Wa.wav 0 days 00:00:00        0 days 00:00:01        -0.000284
                0 days 00:00:00.500000 0 days 00:00:01.500000 -0.000318
```

can be simplified to:

```python
...
interface = audinterface.Feature(
    'mean',
    process_func=mean,
    process_func_applies_sliding_window=True,
    win_dur=win_dur,
    hop_dur=hop_dur,
)
...
```

At least there is a way to avoid setting `win_dur` and `hop_dur` twice.

I would then recommend that we immediately deprecate the default value of `process_func_applies_sliding_window` from `True` to `False`.

File: tests/test_feature.py [hunk 2]
```diff
@@ -847,78 +870,93 @@ def test_process_index(tmpdir):
...
-            marks=pytest.mark.xfail(raises=RuntimeError),
-        ),
+        (0.5, 0.25),
+        (0.5, 0.5),
+        (0.25, 0.5),
+        (pd.to_timedelta(1, unit='s'), pd.to_timedelta(0.5, unit='s')),
+        ('4000', '2000'),
+        ('500ms', '250ms'),
+        ('500milliseconds', '250milliseconds'),
+        (f'{SAMPLING_RATE // 2}', f'{SAMPLING_RATE // 4}'),
```
Comment (hagenw) [inline tests/test_feature.py:892]: In the tests before we also test `hop_dur=None`, did you omit it on purpose now?

File: tests/test_feature.py (hunk 1 above)
Comment (hagenw) [reply tests/test_feature.py:946]: > What's also unfortunate is that - again for backward compatibility - we have to assume by default that the processing function will apply the sliding window, so the default behavior is different from what we have in Process.

In `Process` we don't have the `process_func_applies_sliding_window`, so I guess we don't need to have this argument there and the process func can apply a sliding window or not if we set `win_dur=None`?

Comment (frankenjoe) [reply tests/test_feature.py:946]: In `Process` such an argument would not make sense, since we do assume that the returned values have a certain shape, i.e. we cannot know which axis should be expand into frames.

File: tests/test_feature.py (hunk 2 above)
Comment (frankenjoe) [reply tests/test_feature.py:893]: Oh right, I re-added the test.

File: audinterface/core/feature.py [hunk 3]
```diff
@@ -21,18 +22,35 @@
...
+    the returned object has ``num_channels * num_features`` columns.
     It will have one row per file or signal.
+
     If features are extracted using a sliding window,
     each window will be stored as one row.
     If ``win_dur`` is specified ``start`` and ``end`` indices
     are referred from the original ``start`` and ``end`` arguments
     and the window positions.
     Otherwise, the original ``start`` and ``end`` indices
     are kept.
```
[the remark this answers is not in the document]
Comment (frankenjoe) [reply audinterface/core/feature.py]: Changed to:

image
